frontend/src/Onboarding/TransferCoursesScreen.tsx

```diff
@@ -35,253 +30,159 @@ const TitleText = styled.div`
-  constructor(props: Props) {
-    super(props);
-
-    this.state = {
-      modalVisible: false,
-      selectedRequirements: [],
-      otherCourses: [],
-      completedNonTransfer: props.completedRequirements,
-    };
-  }
-
-  hideModal() {
-    this.setState({ modalVisible: false });
-  }
-
-  showModal() {
-    this.setState({ modalVisible: true });
-  }
-
-  // Adds the "other courses" to the state in the form of a IRequiredCourse[][] so that they can be
-  // processed by renderAllCourses
-  addOtherCourses(courses: ScheduleCourse[]) {
+} & RouteComponentProps;
+
+const coursesToString = (c: IRequiredCourse[]) =>
+  c.map(courseToString).join(",");
+const mappableFetchCourse = (c: IRequiredCourse) =>
+  fetchCourse(c.subject, String(c.classId));
+const fetchAndFilterCourses = async (cs: IRequiredCourse[]) => {
+  const courses = await Promise.all(cs.map(mappableFetchCourse));
+  // filter will get rid of null values, so cast is safe
+  return courses.filter(c => c !== null) as ScheduleCourse[];
+};
+
+const courseEq = (c1: IRequiredCourse, c2: IRequiredCourse) =>
+  c1.classId === c2.classId && c1.subject === c2.subject;
+
+const TransferCoursesScreenComponent: React.FC<TransferCoursesScreenProps> = () => {
+  const major = useSelector((state: AppState) => getUserMajorFromState(state)!);
+  const completedRequirements = useSelector((state: AppState) =>
+    getCompletedRequirementsFromState(state)
+  );
+  const dispatch = useDispatch();
+  const setCompletedCourses = (completedCourses: ScheduleCourse[]) =>
+    dispatch(setCompletedCoursesAction(completedCourses));
+  const setTransferCourses = (transferCourses: ScheduleCourse[]) =>
+    dispatch(setTransferCoursesAction(transferCourses));
+
+  const [modalVisible, setModalVisible] = useState(false);
+  const [selectedRequirements, setSelectedRequirements] = useState<
+    IRequiredCourse[]
+  >([]);
+  const [otherCourses, setOtherCourses] = useState<IRequiredCourse[][]>([]);
+  const [completedNonTransfer] = useState(completedRequirements);
+
+  const addOtherCourses = (courses: ScheduleCourse[]) => {
     let reqCourseMap = courses.map((course: ScheduleCourse) => [
       {
         type: "COURSE",
         classId: +course.classId,
         subject: course.subject,
       } as IRequiredCourse,
     ]);
-    this.setState({
-      otherCourses: [...this.state.otherCourses, ...reqCourseMap],
-    });
-  }
+    setOtherCourses([...otherCourses, ...reqCourseMap]);
+  };
 
   /**
    * Separate transfer and non transfer courses, convert non transfer completed courses
    * into ScheduleCourses, and populate redux store.
    */
-  async onSubmit() {
-    let completedNonTransfer = this.state.completedNonTransfer.filter(
-      req => !this.state.selectedRequirements.includes(req)
+  const onSubmit = async () => {
+    const filteredCompleted = completedNonTransfer.filter(
+      req => !selectedRequirements.some(c => courseEq(c, req))
     );
```

Comment:
this is what I think was the first bug: `includes` uses ref equality (bad). it may have been fine, but it's generally an antipattern to use ref equality.

Reply:
good catch!!